### ISSUE TITLE
`gradle clean build` can run clean after tasks that produce output by depending on other tasks

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#2488]
If a `clean` task depends on other tasks, `clean` can be reordered and made to run after the tasks that would produce output.  This can lead to build failures or just strangeness with seemingly straightforward builds (e.g., `gradle clean build`).

### Expected Behavior

`clean` would respect "destroyable" information and command-line order.

### Current Behavior

When `clean` depends on another task (the example below uses "clean" from another project, but it can be any task), we seem to reorder the clean task in a way that allows it to run after tasks that produce output and in ways that appear to be the opposite of the user's intention (e.g., we appear to be running `gradle build clean` instead of `gradle clean build`).

### Context

This originally showed up in composite build integration tests, but this occurs in regular multi-project builds too.

### Steps to Reproduce (for bugs)
Given this `build.gradle`:
```
allprojects {
    apply plugin: 'java'
}

clean {
    dependsOn ":sub:clean"
}

dependencies {
    compile project(":sub")
}
```

And `settings.gradle`:
```
include 'sub'
```
Running `gradle clean build` will run `:clean` after everything has been built.

Example output:

> :sub:clean
> :processResources NO-SOURCE
> :processTestResources NO-SOURCE
> :sub:compileJava NO-SOURCE
> :sub:processResources NO-SOURCE
> :sub:classes UP-TO-DATE
> :sub:jar
> :compileJava
> :sub:assemble
> :compileJava NO-SOURCE
> :sub:compileTestJava
> :classes UP-TO-DATE
> :sub:compileTestJava NO-SOURCE
> :jar
> :sub:processTestResources NO-SOURCE
> :sub:testClasses UP-TO-DATE
> :sub:test NO-SOURCE
> :sub:check UP-TO-DATE
> :sub:build
> :assemble
> :compileTestJava NO-SOURCE
> :testClasses UP-TO-DATE
> :clean
> :test NO-SOURCE
> :check UP-TO-DATE
> :build

**Comment by ghale** [comment]
Yuck.  I knew this reliance on command line ordering would create a problem somewhere.  

> (the example below uses "clean" from another project, but it can be any task)

Do you mean "any task" as in another task in the same project?  If so, can you provide an example?  Running a few tests myself, I could only get this to fail on multiproject/composite builds.  If there is a failure scenario where the task dependency is in the same project, then this would be very concerning.

What's happening is that we are constructing the task queue according to the order specified on the command line - first, clean and his dependencies (":sub:clean, :clean"), then build and his dependencies (":processResources, :processTestResources, etc").  The first task executor finds ":sub:clean" and executes it, the next executor looks at ":clean" but can't run it because it has dependencies, then he finds ":processResources" which is in another project, and has no relationship to ":sub:clean" (or his destroyables) so he _can_ be selected.  This then starts the producer tasks in ":" and so ":clean" gets delayed until all of those have completed.

The problem of course is the implied intention when running "clean build" which is that all clean tasks should run first and the fact that this order has no significance once the task graph is created (i.e. task relationships trump all).  At that point, we'll simply execute the graph in the quickest order possible.  We are doing it "safely" meaning we don't run ":clean" somewhere in between all of those producer tasks in ":", but it's still different from the implied intention of the command line order.

**Comment by big-guy** [comment]
> Do you mean "any task" as in another task in the same project?

I just meant any task in another subproject. e.g.

```
allprojects {
    apply plugin: 'java'
}

project(":sub") {
    task someTask
}

clean {
    dependsOn ":sub:someTask"
}

dependencies {
    compile project(":sub")
}
```

**Comment by oehme** [comment]
I wonder if we should make the "intended order" a stronger part of the task graph, i.e. introduce "should run after" relationships if they don't conflict with the existing dependencies.

**Comment by idlsoft** [comment]
I'm actually not sure if the tasks given from command line need to form one single graph.
I think the expectation is that
`./gradlew clean build`
is equivalent to
`./gradlew clean`
`./gradlew build`
In other words they are expected to run sequentially, and just skip over uptodate tasks.

**Comment by grv87** [comment]
@idlsoft, I don't agree with you.
Why we should wait running `./gradlew --parallel build1 build2`?

**Comment by mkmaier** [comment]
I think most of the time there is no 'intended order'. My usual expectation is that gradle will execute the required tasks correctly (as defined by task dependencies) and as quickly as possible, nothing more. But in this case, we run into an implicit assumption that 'clean' is special and will always run before any other tasks. If that could be ensured, the problem at hand would be fixed without significant drawbacks to users.
In essence `tasks.* mustRunAfter clean`

**Comment by idlsoft** [comment]
The expectation of tasks executing sequentially in the specified order is very reasonable and based on previous experience with just about any build tool: `make`, `ant`, `maven`, even `gradle`.
The example given by @grv87 is different but it explicitly says `--parallel` to make the intention clear.

I understand that `--parallel` is on by default but I would distinguish between applying it to the execution graph vs the tasks specified from command line. There are going to be other cases, besides `clean` where that order is significant to the user but direct dependencies between tasks are not defined.

And if you do want parallel task execution you'd be able to be explicit about it.

This way it'll work by default always, and faster when you bother to give additional hints.

**Comment by mkmaier** [comment]
@idlsoft, you're only talking about corner cases here where task ordering is not specified, but significant. Personally, I'd consider such behaviour a bug in the build logic, but you might have a different opinion. By extension, `gradle clean build` and `gradle build clean` should do the same thing, namely a clean build.

FWIW, this is what I am using now:

    allprojects {
        project.afterEvaluate {
            def deleteTasks = tasks.withType(Delete) + project.tasks.findByPath('clean')
            def otherTasks = tasks - deleteTasks
            otherTasks*.mustRunAfter deleteTasks
        }
    }
The assumption here is that a `clean` task in project1 does not affect project2 production tasks, but that seems reasonable enough.

**Comment by rocketraman** [comment]
Another case where this fails to work as expected seems to be with composite builds.

**Comment by grv87** [comment]
Maybe Gradle could have a feature to say that some task (for example, `clean`) should run before all other tasks. Excepting tasks that `clean` depends on or explicitly must/should run after.
Then build script could contain something like this:
```
clean.shouldRunFirst()
```
Or it could be built-in behavior for `clean`.

Right now it is not easily feasible. See naive implementation in #6464 and implementation respecting indirect dependencies [here](https://github.com/FIDATA/gradle-prerequisites-plugin/blob/master/src/main/groovy/org/fidata/gradle/PrerequisitesPlugin.groovy#L92-L110).

**Comment by jonathanlocke** [comment]
This just bit me badly. I think "should run first" is too ambiguous. The issue is the same as with hooks or callbacks: who wins if two tasks want to run first? Gradle should have a good default and allow the gradler to alter this order in one place, to make it clear:

executionOrder {
    myBeforeEverythingTask
    clean
    tasksByDependency        // The normal gradle tasks as ordered by dependency
    myAfterEverythingTask
}

I'm new to groovy, so i'm sure that's not the right syntax...

    Jon

**Comment by bartek-wesolowski** [comment]
It would be nice to update the documentation, because it says that tasks are executed in the order they're specified:
https://docs.gradle.org/current/userguide/command_line_interface.html#executing_multiple_tasks
